### PR TITLE
[WIP] 721/1155 refactor

### PR DESCRIPTION
**Pull request by mzhu25**
## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/zero-ex/contracts/src/features/libs/LibNFTOrder.sol [hunk 1]
```diff
@@ -140,6 +191,194 @@ library LibERC1155Order {
...
+        internal
+        pure
+        returns (ERC721Order memory erc721Order)
+    {
+        assembly {
+            erc721Order := nftOrder
+        }
+    }
+
+    function asERC1155Order(
```
Comment (mzhu25) [inline contracts/zero-ex/contracts/src/features/libs/LibNFTOrder.sol:224]: this is kinda expensive but I'm not to concerned about gas costs for 1155

File: contracts/zero-ex/contracts/src/vendor/ITakerCallback.sol [hunk 2]
```diff
@@ -21,14 +21,20 @@ pragma solidity ^0.6.5;
...
+    ///      the taker -> maker transfer.
+    /// @param orderHash The hash of the order being filled when this
+    ///        callback is invoked.
     /// @param callbackData Arbitrary data used by this callback.
-    /// @return success The selector of this function (0x0ea5ba79),
+    /// @return success The selector of this function,
     ///         indicating that the callback succeeded.
-    function zeroExERC1155OrderCallback(bytes calldata callbackData)
+    function zeroExTakerCallback(
+        bytes32 orderHash,
```
Comment (mzhu25) [inline contracts/zero-ex/contracts/src/vendor/ITakerCallback.sol:35]: Consolidated the callbacks and threw in `orderHash` as a parameter as a security measure for integrators

File: contracts/zero-ex/contracts/src/features/nft_orders/NFTOrders.sol [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+                    orderAmount,
+                    fee.amount
+                );
+            }
+            if (feeFillAmount == 0) {
+                continue;
+            }
+
+            if (useNativeToken) {
+                assert(payer == address(this));
```
Comment (dorothy-zbornak) [inline contracts/zero-ex/contracts/src/features/nft_orders/NFTOrders.sol:442]: Maybe also `assert(fee.erc20Token == WETH || fee.erc20Token == ETH)`? :speak_no_evil:

**Review by dorothy-zbornak — COMMENTED** [review]
This code is beautiful in a way only a frustrated solidity developer could appreciate. It's the Rami Malek of smart contracts. It's ugly hot.

File: contracts/zero-ex/contracts/src/features/nft_orders/ERC1155OrdersFeature.sol [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+        bytes32 orderHash,
+        uint128 fillAmount
+    )
+        internal
+        override
+    {
+        LibERC1155OrdersStorage.Storage storage stor = LibERC1155OrdersStorage.getStorage();
+        uint128 currentFilledAmount = uint128(stor.orderState[orderHash]);
+        assert(currentFilledAmount + fillAmount > currentFilledAmount);
+        stor.orderState[orderHash] = currentFilledAmount + fillAmount;
```
[the remark this answers is not in the document]
Comment (mzhu25) [reply contracts/zero-ex/contracts/src/features/nft_orders/ERC1155OrdersFeature.sol:505]: 🙈  yikes good catch

[the remark this answers is not in the document]
Comment (mzhu25) [reply contracts/zero-ex/contracts/src/features/nft_orders/ERC1155OrdersFeature.sol:505]: well I guess it doesn't really matter much, other than for orderInfo
will fix anyway

File: contracts/zero-ex/contracts/src/features/nft_orders/NFTOrders.sol [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+    function _sellNFT(
+        LibNFTOrder.NFTOrder memory buyOrder,
+        uint128 orderAmount,
+        LibSignature.Signature memory signature,
+        SellParams memory params
+    )
+        internal
+        returns (uint256 erc20FillAmount)
+    {
+        LibNFTOrder.OrderInfo memory orderInfo = _getOrderInfo(buyOrder, orderAmount);
```
[the remark this answers is not in the document]
Comment (mzhu25) [reply contracts/zero-ex/contracts/src/features/nft_orders/NFTOrders.sol:88]: I'm confused

[the remark this answers is not in the document]
Comment (mzhu25) [reply contracts/zero-ex/contracts/src/features/nft_orders/NFTOrders.sol:88]: oh wait I think I see what you mean

[the remark this answers is not in the document]
Comment (mzhu25) [reply contracts/zero-ex/contracts/src/features/nft_orders/NFTOrders.sol:88]: I was worried that the `erc1155TokenAmount` word would somehow by the time we cast it back to an `ERC1155Order`, but thinking about it some more I guess it shouldn't happen. Still kinda scary though.

[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply contracts/zero-ex/contracts/src/features/nft_orders/NFTOrders.sol:88]: Wouldn't it just hash wrong and fail the signature check if that happened?

[the remark this answers is not in the document]
Comment (mzhu25) [reply contracts/zero-ex/contracts/src/features/nft_orders/NFTOrders.sol:88]: Oh true, hadn't considered that

**Comment by mzhu25** [comment]
supserseded by https://github.com/0xProject/protocol/pull/429

**Pull request closed without merging**